### ISSUE TITLE
Reserved keywords collision in Yul assembly

### ISSUE THREAD

**Issue by shekhirin** [hyperledger-labs/solang#639]
Such snippet
```solidity
contract Contract {
    constructor() {
        assembly {
            return(0, 0)
        }
    }
}
```
can't be parsed due to 
```
 [Diagnostic { level: Error, ty: ParserError, pos: Some(Loc(0, 71, 77)), message: "unrecognised token `return', expected \"address\", \"bool\", \"break\", \"bytes\", \"continue\", \"false\", \"for\", \"if\", \"leave\", \"let\", \"mapping\", \"payable\", \"string\", \"switch\", \"true\", \"}\", Bytes, Int, Uint, hexnumber, identifier, number, string", notes: [] }]
```
error. The reason for this is that `return` is a reserved keyword in Solidity but just a regular function in Yul.

This is an issue similar to https://github.com/hyperledger-labs/solang/pull/636#pullrequestreview-853224787, i.e. requires separation of Yul and Solidity keywords to prevent such collisions.

**Comment by seanyoung** [comment]
So I have been thinking about this problem for some time. I would like it to be possible to specify amounts like `1 sol` or `50000 lamports` for Solana, without introducing new keywords. 

The lalrpop parser has a separate lexer, so this is hard to work around. One way is not make `return` a keyword, and then parse `return ...` as `Identifier ...` and then check if the identifier is `return`. This is how you traditionally work around these issues, but it's a workaround.

Another issue is #625. I don't think this is possible to fix without some changes to lalrpop itself.

So I have been thinking about this problem for a while, and I've been reading a lot about parser generators. My first idea was to use peg grammars, which allow for left recursion. This is why I started https://github.com/seanyoung/lrpeg This would solve the keyword issue.

However, peg grammars don't have great performance when it comes to back-tracking. Rat-pack parsing helps a little, but LR type grammars are much faster, as much less backtracking is done.

There is another issue with peg grammars and lalrpop. You want the parser to parse as much as possible when syntax errors are present. In other words, if someone types "for let ()" in one function, parsing should not fail completely. This is important for code editors / language servers, like the solang vscode extension. Code is mostly broken when the developer is typing.

So I think we want GLR parsing, with error recovery and context-sensitive keyword parsing. This has lead me to consider tree sitter, which has all these properties. There is also a grammar for solidity https://crates.io/crates/tree-sitter-solidity

I had not considered switching parsers for solang necessary yet, but now we starting to find issues with parsing yul, so this is much more urgent. 

@shekhirin I know you have invested quite some time solang-parser, however how do you feel about trying tree-sitter-solidity instead? We can combine our efforts to make sure this parser is complete, you for your code formatter and me for the solang compiler.

**Comment by shekhirin** [comment]
@seanyoung thank you for the extensive explanation!

> This is how you traditionally work around these issues, but it's a workaround.

I think that **for us and formatter purposes** having a separate parsing pipeline for `return` keyword is a better workaround than using other parser such as `tree-sitter`. I understand however that you might not be willing to add this since parsing `assembly` isn't critical for your goals at all. I think I'll look into it and figure out whether it's a hard thing to change or not and maybe we could just have a fork of solang for formatter purposes.

> Another issue is #625. I don't think this is possible to fix without some changes to lalrpop itself.

I think it's possible if we'll add `Comment` token to all places we can encounter it: https://github.com/hyperledger-labs/solang/pull/640. That seems to be a working solution and imo rather clean but it needs to be well-tested with all possible places where comment can appear.

**Comment by seanyoung** [comment]
> @seanyoung thank you for the extensive explanation!
> 
> > This is how you traditionally work around these issues, but it's a workaround.
> 
> I think that **for us and formatter purposes** having a separate parsing pipeline for `return` keyword is a better workaround than using other parser such as `tree-sitter`. I understand however that you might not be willing to add this since parsing Yul assembly isn't critical for your goals at all. I think I'll look into it and figure out whether it's a hard thing to change or not and maybe we could just have a fork of solang for formatter purposes.

On the contrary, yul assembly is very important for solang.

> > Another issue is #625. I don't think this is possible to fix without some changes to lalrpop itself.
> 
> I think it's possible if we'll add `Comment` token to all places we can encounter it: #640. That seems to be a working solution and imo rather clean but it needs to be well-tested with all possible places where comment can appear.

That's an interesting idea.

Ok, let's try and make this work, that's fine. In the mean time, I'll be investigating tree-sitter as well.

**Comment by seanyoung** [comment]
I'm working on this at the moment, I'm trying a few different ways of doing this. May take a few days

**Comment by seanyoung** [comment]
This is turning out to be quite painful.

**Comment by seanyoung** [comment]
After all of that I had a rather obvious insight this morning. We can fix it with the existing parser generator.

**Comment by seanyoung** [comment]
Fixed in #652
